lib/platform/linux-uart.c

```diff
@@ -0,0 +1,423 @@
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(1000)
+
+
+static int read_resp_line(int fd, char *str)
+{
+	int ret;
+	int i;
+	int cnt = 0;
+
+	for (i = 0; i < RETRY_NUM; i++) {
+
+		ret = read(fd, str + cnt, sizeof(str));
+		if (ret < 0)
+			return ret;
+	
+		cnt += ret;
+		str[cnt] = '\0';
+	
+		if (strrchr(str, '>') && strrchr(str, ':') &&
+			(strrchr(str, ':') + 5 == strrchr(str, '>'))){
+			return 0;
+		}
+	}
+	
+	return -1;
+}
+
+static int cli_control(struct switchtec_dev *dev, const char *str)
+{
+	int ret;
+	char rtn[1024];
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+
+	ret = write(udev->fd, str, strlen(str));
+	if (ret != strlen(str))
+		return -1;
+
+	ret =  read_resp_line(udev->fd, rtn);
+	if (ret)
+		return ret;
+		
+	
+	return 0;
+}
+
+#ifdef __CHECKER__
+#define __force __attribute__((force))
+#else
+#define __force
+#endif
+
+static void uart_close(struct switchtec_dev *dev)
+{
+	struct switchtec_uart *udev =  to_switchtec_uart(dev);
+	cli_control(dev, echo_on);
+
+	if (dev->gas_map)
+		munmap((void __force *)dev->gas_map, dev->gas_map_size);
+
+	flock(udev->fd, LOCK_UN);
+	close(udev->fd);
+	free(udev);
+}
+
+static int map_gas(struct switchtec_dev *dev)
+{
+	void *addr;
+	dev->gas_map_size = 4 << 20;
+
+	addr = mmap(NULL, dev->gas_map_size, PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (addr == MAP_FAILED)
+		return -1;
+
+	dev->gas_map = (gasptr_t __force)addr;
+
+	return 0;
+}
+
+#undef __force
+
+static gasptr_t uart_gas_map(struct switchtec_dev *dev, int writeable, size_t *map_size)
+{
+	if (map_size)
+		*map_size = dev->gas_map_size;
+
+	return dev->gas_map;
+}
+
+
+static void uart_memcpy_from_gas(struct switchtec_dev *dev, void *dest, const void __gas *src, size_t n)
+{
+	int ret, raddr, rnum, rcrc, idx;
+	int i;
+	int crc_rty_cnt = 0;
+	char *tok;
+	const char *delims = "]";
+	char gas_rd[1024];
+	char gas_rd_rtn[102400];
+	char *b = gas_rd;
+	uint8_t crc_cal;
+	uint8_t crc_buf[2*n];
+	uint8_t *crc_ptr = crc_buf;
+	uint32_t crc_cnt;
+	uint32_t crc_addr;
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+	uint32_t addr = (uint32_t)(src - (void __gas *)dev->gas_map);
+
+	b += snprintf(b, sizeof(gas_rd), "gasrd -c -s 0x%x %zu\r", addr, n);
+
+rd_crc_err:
+	if (crc_rty_cnt > RETRY_NUM)
+		exit(1);
+
+	ret = write(udev->fd, gas_rd, strlen(gas_rd));
+	if (ret != strlen(gas_rd))
+		exit(1);
+
+	ret = read_resp_line(udev->fd, gas_rd_rtn);
+	if (ret)
+		exit(1);
+
+	if(sscanf(gas_rd_rtn, "%*[^<]<0x%x> [%d Byte]%*[^:]: 0x%x%*[^:]:%x>",
+			  &raddr, &rnum, &rcrc, &idx) != 4)
+		exit(1);
+
+	assert(raddr == addr);
+	assert(rnum == n);
+
+	crc_addr = htobe32(addr);
+	memcpy(crc_ptr, &crc_addr, sizeof(crc_addr));
+	crc_ptr += sizeof(addr);
+
+	tok = strtok(gas_rd_rtn,delims);
+	for (i = 0; i<rnum; i++){
+		tok = strtok(NULL, " ");
+		*crc_ptr = strtol(tok, &tok, 16);
```

Comment:
This is confusing. You overwrite tok with the endptr in strtol, but it's never used again (strtok does not use the value of tok to continue parsing on, it uses a stored value).

It also seems like strtok isn't really useful here....

Use strchr to find the first "[' then use strtol to parse each number (it ignores spaces and returns the end of the parsed value).